### PR TITLE
Upgrade React to 15.4

### PR DESCRIPTION
**Pull request by singingwolfboy**
As well as a few other JS packages related to React. [Changelog is here.](https://github.com/facebook/react/blob/master/CHANGELOG.md)

### CODE REVIEW

**Comment by pdpinch** [comment]
Is there any reason to do this upgrade? 

How are we going to test it?

**Comment by singingwolfboy** [comment]
@pdpinch No special reason beyond keeping up with the upgrade treadmill. Frequent small upgrades are easier and less risky than infrequent large upgrades. Just as we want to use the latest version of Django, we should also use the latest version of React.

**Comment by aliceriot** [comment]
the release notes are [here](https://facebook.github.io/react/blog/2016/11/16/react-v15.4.0.html) for 15.4, there shouldn't be any breaking changes unless we've depended somewhere on implementation details / library internals (we haven't).

**Comment by aliceriot** [comment]
Can you make sure that the new packages are saved correctly in `npm-shrinkwrap.json`?

If you're not familiar with shrinkwrap doing something like the following should work:

```
git checkout master -- package.json
npm install
npm install --save react@15.4.0 react-dom@15.4.0 eslint-plugin...
```

if all goes well the changes will be reflected in the shrinkwrap file when you do that.

If that doesn't work there could be a few things going on. Often things don't get saved automatically in it because the `node_modules` can get out of sync with it, and then `npm` doesn't know what to do. `npm prune` can be helpful in dealing with that situation.

**Comment by aliceriot** [comment]
( the reason it's important is that the shrinkwrap file overrules package.json in terms of specifying semver )

**Comment by singingwolfboy** [comment]
Fixed. Thanks for the explanation and guide!

**Comment by singingwolfboy** [comment]
Looks like this upgrade caused a few tests to fail, and exposed ways that we're using React incorrectly. Can someone else look into fixing those problems? Someone else who is more familiar with our codebase could probably do a better job than I could.

**Comment by aliceriot** [comment]
I'm going to stage a friendly takeover of this PR! haha

**Comment by singingwolfboy** [comment]
I, for one, welcome our new Alice overlords.

alice overlord

(GIF is unrelated, but awesome.)

**Comment by aliceriot** [comment]
haha

looks like some of the issues are from material-ui. there's a PR open there for a release which should fix it: https://github.com/callemall/material-ui/pull/5580 I think we probs need to wait for that to be merged to make much progress.

**Comment by aliceriot** [comment]
@noisecapella Hey, this is ready for review if you're free.

There is currently a `PropTypes` related warning on `LearnerSearchPage_test.js` which is related to dependencies of Searchkit. I'm going to file an issue to see if they're willing to work on react 15.4 compatibility (especially because these warnings will become errors in react 16).

Anyway, the react documentation relating to the warning is [here](https://facebook.github.io/react/warnings/dont-call-proptypes.html).

**Comment by aliceriot** [comment]
But w/r/t those warnings I'm not sure how to go about getting rid of them otherwise. I tried aliasing some of the dependencies of searchkit, but I couldn't get that working.

**Comment by aliceriot** [comment]
@noisecapella this is fairly WIP right now, I had to do a bunch of stuff to get webpack hot module reloading working again and it seems like maybe it doesn't work properly with react-router

**Comment by aliceriot** [comment]
I got the hot module reloading working! yay.

I think I have some linting errors and so on to fix but other than that I think this is ready for review

File: webpack_dev_server.sh [hunk 1]
```diff
@@ -35,6 +35,8 @@ elif [[ -z "$WEBPACK_DEV_SERVER_HOST" || "$IS_OSX_HOST_MACHINE" == 'true' ]] ; t
     npm install && echo "Finished npm install"
   fi
   # Start the webpack dev server on the appropriate host and port
-  node ./node_modules/webpack-dev-server/bin/webpack-dev-server.js --config webpack.config.dev.js -d \
-       --content-base ./static --host "$WEBPACK_HOST" --port "$WEBPACK_PORT" --progress --inline --hot
+  # node ./node_modules/webpack-dev-server/bin/webpack-dev-server.js --config webpack.config.dev.js -d \
```
Comment (noisecapella) [inline webpack_dev_server.sh]: Can you remove the commented out code?

Comment (aliceriot) [inline webpack_dev_server.sh]: oh yea oops

File: .eslintrc [hunk 2]
```diff
@@ -21,6 +21,7 @@
         "location"
       ]}],
     "react/display-name" : [0],
+    "react/no-find-dom-node": [0],
```
Comment (noisecapella) [inline .eslintrc]: I remember back when `getDOMNode` was deprecated and `ReactDOM.findDOMNode` was the new thing :frowning_face:

File: static/js/components/search/CustomSortingSelect.js [hunk 3]
```diff
@@ -25,7 +25,7 @@ export default class CustomSortingSelect extends Select {
         text += ` (${countFormatter(item.docCount)})`;
       }
 
-      return <option key={item.key} value={item.key} disable={item.disabled}>
+      return <option key={item.key} value={item.key}>
```
Comment (noisecapella) [inline static/js/components/search/CustomSortingSelect.js]: What does this change mean?

File: static/js/entry/dashboard.js [hunk 4]
```diff
@@ -1,11 +1,14 @@
+require('react-hot-loader/patch');
```
Comment (noisecapella) [inline static/js/entry/dashboard.js]: What does this do?

File: .eslintrc (hunk 2 above)
Comment (aliceriot) [inline .eslintrc]: yeah... the JS constant API churn is a bit annoying.

File: static/js/entry/dashboard.js (hunk 4 above)
Comment (aliceriot) [inline static/js/entry/dashboard.js]: This ensures that the code that bootstraps the react-hot-loader is executed before anything else. it spews a warning otherwise.

File: static/js/util/test_utils.js [hunk 5]
```diff
@@ -80,7 +80,10 @@ export const modifySelectField = (field: HTMLElement, text: string): void => {
 };
 
 export const clearSelectField = (field: HTMLElement): void => {
-  TestUtils.Simulate.mouseDown(field.querySelector('.Select-clear'), { button: 1 });
+  let input = field.querySelector('.Select-input').querySelector('input');
```
Comment (noisecapella) [inline static/js/util/test_utils.js]: Is this related to @roberthouse54's change to remove the 'x' for select fields?

File: static/js/entry/dashboard.js [hunk 6]
```diff
@@ -25,11 +28,30 @@ const store = configureStore();
...
 
+const rootEl = document.getElementById("dashboard");
 ReactDOM.render(
-  makeDashboardRoutes(
-    browserHistory,
-    store,
-    () => ga.pageview(window.location.pathname)
-  ),
-  document.getElementById("dashboard")
+  <AppContainer>
```
Comment (aliceriot) [inline static/js/entry/dashboard.js]: Yeah, you can read about how it works here: https://github.com/gaearon/redux-devtools/commit/64f58b7010a1b2a71ad16716eb37ac1031f93915#diff-363d10e8c7d704f50ac23a129f4a123cR11

Comment (aliceriot) [inline static/js/entry/dashboard.js]: Basically it's a no-op in production.

File: static/js/components/search/CustomSortingSelect.js (hunk 3 above)
Comment (noisecapella) [inline static/js/components/search/CustomSortingSelect.js]: What about changing it to `disabled` instead?

File: static/js/entry/dashboard.js [hunk 7]
```diff
@@ -25,11 +28,30 @@ const store = configureStore();
...
+    />
+  </AppContainer>,
+  rootEl
 );
+
+if (module.hot) {
+  module.hot.accept('../DashboardRouter', () => {
+    const DashboardRouterNext = require('../DashboardRouter').default;
+    ReactDOM.render(
+      <AppContainer>
```
Comment (noisecapella) [inline static/js/entry/dashboard.js]: Can we move this into a variable and use it above too?

Comment (aliceriot) [inline static/js/entry/dashboard.js]: Sure

File: static/js/util/test_utils.js (hunk 5 above)
Comment (noisecapella) [inline static/js/util/test_utils.js]: Was there something in the old code which was breaking under the new libraries?

Comment (aliceriot) [inline static/js/util/test_utils.js]: Yes, one of the tests for the `FinancialAidCalculator`.

File: static/js/components/search/CustomSortingSelect.js (hunk 3 above)
Comment (aliceriot) [inline static/js/components/search/CustomSortingSelect.js]: oh yup, that works!

File: .eslintrc (hunk 2 above)
Comment (noisecapella) [inline .eslintrc]: It looks like we use `findDOMNode` in only two places. Do you think it's worth the effort to fix this now?

File: static/js/DashboardRouter.js [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    } = this.props;
+    const username = SETTINGS.user ? SETTINGS.user.username : '';
+
+    return <div>
+      <MuiThemeProvider muiTheme={getMuiTheme()}>
+        <Provider store={store}>
+          <Router history={browserHistory} onUpdate={onRouteUpdate}>
+            <Route path="/" component={App}>
+              <Route path="dashboard" component={DashboardPage} />
+              <Route path="profile" component={ProfilePage} />
```
Comment (aliceriot) [inline static/js/DashboardRouter.js]: sorry ;_;

**Comment by noisecapella** [comment]
Functionality looks good on local instance, looking at PR build...

**Comment by aliceriot** [comment]
I think we need someone who uses a Mac to make sure everything is working there too, @giocalitri or @gsidebo, would either of you be able to pull this down and make sure JS stuff is working as normal?

File: .eslintrc (hunk 2 above)
Comment (aliceriot) [inline .eslintrc]: actually the profile page one isn't trivial, not sure what's gong on. if you're cool with it I'd like to leave it off for now.

Comment (noisecapella) [inline .eslintrc]: Sure, can you file an issue for it?

Comment (aliceriot) [inline .eslintrc]: https://github.com/mitodl/micromasters/issues/1813

File: static/js/entry/dashboard.js (hunk 4 above)
Comment (aliceriot) [inline static/js/entry/dashboard.js]: I've only set up the react hot-reload for the dashboard

Comment (noisecapella) [inline static/js/entry/dashboard.js]: It would be nice to have this for the other entry points but we can do that in another PR

**Comment by noisecapella** [comment]
PR build and code looks good, waiting on a Mac review

File: static/js/DashboardRouter.js (hunk 8 above)
Comment (rhysyngsun) [inline static/js/DashboardRouter.js]: no prob, this is comparatively minor to the issues I've seen so far

**Comment by gsidebo** [comment]
looks good on mac 👍

**Comment by aliceriot** [comment]
weee here we go!
